k8s/persistent_volumes_test.go

```diff
@@ -0,0 +1,144 @@
+//go:build integration
+// +build integration
+
+package k8s_test
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"knative.dev/kn-plugin-func/k8s"
+)
+
+func TestUploadToVolume(t *testing.T) {
+	var err error
+	var ctx = context.Background()
+
+	cliSet, err := k8s.NewKubernetesClientset()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	creatOpts := metav1.CreateOptions{}
+	deleteOpts := metav1.DeleteOptions{}
+
+	testingNS := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "volume-uploader-test-ns-" + rand.String(5),
+		},
+		Spec: corev1.NamespaceSpec{},
+	}
+
+	_, err = cliSet.CoreV1().Namespaces().Create(ctx, testingNS, creatOpts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		cliSet.CoreV1().Namespaces().Delete(ctx, testingNS.Name, deleteOpts)
+	})
+	t.Log("created namespace: ", testingNS.Name)
+
+	testingPVCName := "testing-pvc"
+
+	err = k8s.CreatePersistentVolumeClaim(ctx, testingPVCName, testingNS.Name,
+		nil, nil,
+		corev1.ReadWriteOnce, *resource.NewQuantity(1024, resource.DecimalSI))
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Log("created PVC: " + testingPVCName)
+
+	// First, test error handling by uploading empty content stream.
+	err = k8s.UploadToVolume(ctx, &bytes.Buffer{}, testingPVCName, testingNS.Name)
+	if err == nil || !strings.Contains(err.Error(), "short read") {
+		t.Error("got <nil> error, or error with unexpected message")
+	}
+
+	f, err := os.Open(filepath.Join("testData", "content.tar"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { f.Close() })
```

Comment:
I don't think so. The `Close()` returns error whereas cleanup routine is `void`.

But, yes it could work in a language with reasonable function type sub-typing (which `Go` is definitely not):
i.e. `f: A->B` is sub-type of `g: C->D` iff `A super-type of C ∧ B is sub-type of D`.